consul/state/catalog.go

```diff
@@ -662,15 +688,34 @@ func (s *StateStore) NodeServices(ws memdb.WatchSet, nodeName string) (uint64, *
 	// Get the table index.
 	idx := maxIndexTxn(tx, "nodes", "services")
 
-	// Query the node
+	// Query the node by node name
 	watchCh, n, err := tx.FirstWatch("nodes", "id", nodeName)
 	if err != nil {
 		return 0, nil, fmt.Errorf("node lookup failed: %s", err)
 	}
-	ws.Add(watchCh)
+
 	if n == nil {
-		return 0, nil, nil
+		if len(nodeName) >= minUUIDLookupLen {
+			// Attempt to lookup the node by it's node ID
+			var idWatchCh <-chan struct{}
+			idWatchCh, n, err = tx.FirstWatch("nodes", "uuid_prefix", nodeName)
```

Comment:
I think you wan't to use the iterator form here to make sure the lookup isn't ambiguous:

```
iter, err := tx.Get("nodes", "uuid_prefix", nodeName)
n = iter.Next()
if n != nil {
  if iter.Next() == nil {
    // got it
  } else {
    // it was ambiguous, return nil
  }
}
```